### ISSUE TITLE
DrainModeManager is unnecessary

### ISSUE THREAD

**Issue by mathewc** [Azure/azure-webjobs-sdk#2776]
In https://github.com/Azure/azure-webjobs-sdk/pull/2526 a new IDrainModeManager service was added. The goal as described in that issue was to allow an API to be called that would stop the host listeners, and signal cancellation tokens for any in progress invocations. **However, that's precisely what JobHost.StopAsync does already** before these changes (see code [here](https://github.com/Azure/azure-webjobs-sdk/blob/b6d5b52da5d2fb457efbf359cbdd733186aacf7c/src/Microsoft.Azure.WebJobs.Host/JobHost.cs#L126)). In general the JobHost StartAsync/StopAsync start/stop listeners.

So unless I'm missing something the changes were unnecessary. WebJobs SDK customers don't use this drain interface, they just call StopAsync on their host. For Azure Functions scenarios, the functions host is managing ScriptHost instances. When an instance needs to be "drained", StopAsync can simply be called on it in the same way.

In addition to the increased concept count, there's performance/memory overhead in the implementation [here](https://github.com/Azure/azure-webjobs-sdk/blob/62bbc94767672b17600f8329a5d36fee364a7c72/src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs#L253) since we're doing a bunch of invocation book-keeping.

Likely the only consumer of this is Azure Functions. We could change the implementation of the drain service we register to call StopAsync on the host when EnableDrainModeAsync is called. Then we'd remove all the book-keeping and listener stuff mentioned above. We'd mark the types as [Obsolete].

**Comment by liliankasem** [comment]
Hey Mathew! Just want to try and wrap my head around this and see if this is something I can maybe pick up after resume in the next sprint.

You're suggesting we keep `DrainModeManager` but change the implementation so that `EnableDrainModeAsync` calls `JobHost.StopAsync` instead?

1. What happens to `CancelFunctionInvocations();` - I don't see StopAsync doing anything around that, so would we still need to handle this in `EnableDrainModeAsync`? 

2. Regarding the stopping of listeners, we currently loop through all the registered listeners and call `StopAsync` on each one of those. I see that `JobHost.StopAsync` also calls `StopAsync` on a listener but I don't quite understand the difference here - what listener are we stopping and how does that tie in to stopping all of the listeners? I feel like I'm missing something :D 

I think I need to spend time to better understand how the listeners work, how they're started and how they're stopped so I can see how it plays into Drain Mode stopping all of the listeners. If there are any resources/documentation on that, it would be super useful for me to upskill on.

**Comment by mathewc** [comment]
No, I'm saying that none of the drain mode interfaces added were needed. The existing JobHost.StopAsync API is the way clients stop listeners and spin down a host. It handles the stopping of all listeners [here](https://github.com/Azure/azure-webjobs-sdk/blob/bba8d6a69f77f3fea870082efcd540a138a788b4/src/Microsoft.Azure.WebJobs.Host/JobHost.cs#L135). This listener is the aggregate composite listener that composes all listeners. It also cancels the root cancellation token which is linked to all tokens passed into individual function invocations.

To see how the listeners are all chained together, recommend simply debugging through host startup and host stop to see all this in action. Basically there's a single root composite listener created [here](https://github.com/Azure/azure-webjobs-sdk/blob/3f4ec78be9f43bb041937425ced00b341883aa42/src/Microsoft.Azure.WebJobs.Host/Listeners/HostListenerFactory.cs#L89).

**Comment by drdamour** [comment]
@mathewc what would be the equivalent in JobHost stack of taking a DI dependency on IDrainModeManager and looking at IsDrainModeEnabled in my long running functions that early exit when that's set?
